src/main/java/org/commcare/core/network/ModernHttpRequester.java

```diff
@@ -72,26 +82,35 @@ private void setupAuthentication(boolean isAuth, User user, String domain) {
                 username = user.getUsername();
             }
             final String password = user.getCachedPwd();
-            if (username == null || password == null || User.TYPE_DEMO.equals(user.getUserType())) {
-                String message =
-                        "Trying to make authenticated http request without proper credentials";
-                throw new RuntimeException(message);
-            } else if (!"https".equals(url.getProtocol())) {
-                throw new PlainTextPasswordException();
-            } else {
-                Authenticator.setDefault(new Authenticator() {
-                    @Override
-                    protected PasswordAuthentication getPasswordAuthentication() {
-                        return new PasswordAuthentication(username, password.toCharArray());
-                    }
-                });
-            }
+            setupAuthentication(username, password, user);
         } else {
             // clear any prior set authenticator to make unauthed requests
             Authenticator.setDefault(null);
         }
     }
 
+    private void setupAuthentication(final String username, final String password, User user) {
+        if (username == null || password == null ||
+                (user != null && User.TYPE_DEMO.equals(user.getUserType()))) {
```

Comment:
I'm a bit confused what the various paths through the two setupAuthentication methods are - are they:

1. Web user, not logged in
2. Mobile user, not logged in
3. Mobile user, logged in
4. Demo user, logged in

?

Either way feels strange to check whether its a demo user this far along - might be better to not expose this feature to demo users from the outset?